File: lib/features/operations/domain/system_configuration.dart
enum FeatureFlag {
  whatsappNotifications,
  aiRecommendations,
  automatedCollection,
  providerSelfPublication,
  premiumSubscription,
  documentAnalysis,
  newMobileFeatures,
  experimentalSearch,
}

class PlatformConfiguration {
  const PlatformConfiguration({
    required this.version,
    required this.platformName,
    required this.logoLocation,
    required this.brandPrimaryColor,
    required this.emailSenderName,
    required this.emailSenderAddress,
    required this.verificationExpirationDays,
    required this.supportedCountries,
    required this.supportedLanguages,
    required this.opportunityCategories,
    required this.documentTypes,
    required this.maximumFileSizeBytes,
    required this.applicantRegistrationEnabled,
    required this.providerRegistrationEnabled,
    required this.maintenanceMode,
    required this.featureFlags,
    required this.environment,
    required this.securityPolicy,
    required this.recommendationSettings,
    required this.fraudRuleSettings,
    required this.integrationSettings,
    required this.notificationSettings,
    required this.updatedAt,
    required this.updatedBy,
    required this.changeReason,
  });

  factory PlatformConfiguration.defaults() => PlatformConfiguration(
    version: 1,
    platformName: 'ScholarSphere',
    logoLocation: 'assets/branding/logo.png',
    brandPrimaryColor: '#007C72',
    emailSenderName: 'ScholarSphere',
    emailSenderAddress: 'no-reply@scholarsphere.example',
    verificationExpirationDays: 90,
    supportedCountries: const ['Global'],
    supportedLanguages: const ['en'],
    opportunityCategories: const ['scholarship', 'fellowship', 'internship'],
    documentTypes: const ['passport', 'transcript', 'curriculum_vitae'],
    maximumFileSizeBytes: 10 * 1024 * 1024,
    applicantRegistrationEnabled: true,
    providerRegistrationEnabled: true,
    maintenanceMode: false,
    featureFlags: const {
      FeatureFlag.whatsappNotifications: false,
      FeatureFlag.aiRecommendations: false,
      FeatureFlag.automatedCollection: false,
      FeatureFlag.providerSelfPublication: false,
      FeatureFlag.premiumSubscription: false,
      FeatureFlag.documentAnalysis: false,
      FeatureFlag.newMobileFeatures: false,
      FeatureFlag.experimentalSearch: false,
    },
    environment: const {'name': 'development', 'apiVersion': 'v1'},
    securityPolicy: const {
      'minimumPasswordLength': 12,
      'mfaForAdministrators': true,
    },
    recommendationSettings: const {'diversityEnabled': true, 'limit': 20},
    fraudRuleSettings: const {'automaticHideScore': 90},
    integrationSettings: const {'sandboxEnabled': true},
    notificationSettings: const {'dailyLimit': 10},
    updatedAt: null,
    updatedBy: 'system',
    changeReason: 'Initial configuration',
  );

  final int version;
  final String platformName;
  final String logoLocation;
  final String brandPrimaryColor;
  final String emailSenderName;
  final String emailSenderAddress;
  final int verificationExpirationDays;
  final List<String> supportedCountries;
  final List<String> supportedLanguages;
  final List<String> opportunityCategories;
  final List<String> documentTypes;
  final int maximumFileSizeBytes;
  final bool applicantRegistrationEnabled;
  final bool providerRegistrationEnabled;
  final bool maintenanceMode;
  final Map<FeatureFlag, bool> featureFlags;
  final Map<String, Object> environment;
  final Map<String, Object> securityPolicy;
  final Map<String, Object> recommendationSettings;
  final Map<String, Object> fraudRuleSettings;
  final Map<String, Object> integrationSettings;
  final Map<String, Object> notificationSettings;
  final DateTime? updatedAt;
  final String updatedBy;
  final String changeReason;

  bool enabled(FeatureFlag flag) => featureFlags[flag] ?? false;

  PlatformConfiguration copyWith({
    int? version,
    String? platformName,
    String? logoLocation,
    String? brandPrimaryColor,
    String? emailSenderName,
    String? emailSenderAddress,
    int? verificationExpirationDays,
    List<String>? supportedCountries,
    List<String>? supportedLanguages,
    List<String>? opportunityCategories,
    List<String>? documentTypes,
    int? maximumFileSizeBytes,
    bool? applicantRegistrationEnabled,
    bool? providerRegistrationEnabled,
    bool? maintenanceMode,
    Map<FeatureFlag, bool>? featureFlags,
    Map<String, Object>? environment,
    Map<String, Object>? securityPolicy,
    Map<String, Object>? recommendationSettings,
    Map<String, Object>? fraudRuleSettings,
    Map<String, Object>? integrationSettings,
    Map<String, Object>? notificationSettings,
    DateTime? updatedAt,
    String? updatedBy,
    String? changeReason,
  }) => PlatformConfiguration(
    version: version ?? this.version,
    platformName: platformName ?? this.platformName,
    logoLocation: logoLocation ?? this.logoLocation,
    brandPrimaryColor: brandPrimaryColor ?? this.brandPrimaryColor,
    emailSenderName: emailSenderName ?? this.emailSenderName,
    emailSenderAddress: emailSenderAddress ?? this.emailSenderAddress,
    verificationExpirationDays:
        verificationExpirationDays ?? this.verificationExpirationDays,
    supportedCountries: supportedCountries ?? this.supportedCountries,
    supportedLanguages: supportedLanguages ?? this.supportedLanguages,
    opportunityCategories: opportunityCategories ?? this.opportunityCategories,
    documentTypes: documentTypes ?? this.documentTypes,
    maximumFileSizeBytes: maximumFileSizeBytes ?? this.maximumFileSizeBytes,
    applicantRegistrationEnabled:
        applicantRegistrationEnabled ?? this.applicantRegistrationEnabled,
    providerRegistrationEnabled:
        providerRegistrationEnabled ?? this.providerRegistrationEnabled,
    maintenanceMode: maintenanceMode ?? this.maintenanceMode,
    featureFlags: featureFlags ?? this.featureFlags,
    environment: environment ?? this.environment,
    securityPolicy: securityPolicy ?? this.securityPolicy,
    recommendationSettings:
        recommendationSettings ?? this.recommendationSettings,
    fraudRuleSettings: fraudRuleSettings ?? this.fraudRuleSettings,
    integrationSettings: integrationSettings ?? this.integrationSettings,
    notificationSettings: notificationSettings ?? this.notificationSettings,
    updatedAt: updatedAt ?? this.updatedAt,
    updatedBy: updatedBy ?? this.updatedBy,
    changeReason: changeReason ?? this.changeReason,
  );
}

File: scholarsphere_backend/app/api/routes/system_configuration.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth import AuthenticatedUser, get_current_user
from app.core.rbac import require_roles
from app.db.session import get_db
from app.models.audit_log import AuditAction, AuditResult
from app.models.system_configuration import PlatformConfiguration
from app.schemas.system_configuration import (
    DEFAULT_FEATURE_FLAGS,
    PlatformConfigurationRead,
    RollbackRequest,
    SaveConfigurationRequest,
)
from app.services.audit_log import append_audit_record
from app.services.parsing import utc_now

router = APIRouter(prefix="/system-configuration", tags=["system configuration"])

# Matches DemoSystemConfigurationRepository._authorize exactly.
config_access = Depends(require_roles("administrator", "securityAdministrator", "superAdministrator"))


async def _current(session: AsyncSession) -> PlatformConfiguration:
    config = await session.scalar(
        select(PlatformConfiguration).order_by(PlatformConfiguration.version.desc()).limit(1)
    )
    if config is None:
        config = PlatformConfiguration(
            version=1,
            platform_name="ScholarSphere",
            logo_location="assets/branding/logo.png",
            brand_primary_color="#007C72",
            email_sender_name="ScholarSphere",
            email_sender_address="no-reply@scholarsphere.example",
            verification_expiration_days=90,
            supported_countries=["Global"],
            supported_languages=["en"],
            opportunity_categories=["scholarship", "fellowship", "internship"],
            document_types=["passport", "transcript", "curriculum_vitae"],
            maximum_file_size_bytes=10 * 1024 * 1024,
            applicant_registration_enabled=True,
            provider_registration_enabled=True,
            maintenance_mode=False,
            feature_flags=dict(DEFAULT_FEATURE_FLAGS),
            environment={"name": "development", "apiVersion": "v1"},
            security_policy={"minimumPasswordLength": 12, "mfaForAdministrators": True},
            recommendation_settings={"diversityEnabled": True, "limit": 20},
            fraud_rule_settings={"automaticHideScore": 90},
            integration_settings={"sandboxEnabled": True},
            notification_settings={"dailyLimit": 10},
            updated_at=None,
            updated_by="system",
            change_reason="Initial configuration",
        )
        session.add(config)
        await session.flush()
    return config


def _check_cross_field_rules(payload: SaveConfigurationRequest) -> None:
    if payload.feature_flags.get("providerSelfPublication") and not payload.security_policy.get(
        "mfaForAdministrators"
    ):
        raise HTTPException(
            status_code=409,
            detail="Provider self-publication requires the stronger security policy.",
        )


@router.get("/current", response_model=PlatformConfigurationRead)
async def current_configuration(
    _: Annotated[AuthenticatedUser, config_access],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> PlatformConfigurationRead:
    return PlatformConfigurationRead.model_validate(await _current(session))


@router.get("/history", response_model=list[PlatformConfigurationRead])
async def configuration_history(
    _: Annotated[AuthenticatedUser, config_access],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> list[PlatformConfigurationRead]:
    await _current(session)
    rows = await session.scalars(
        select(PlatformConfiguration).order_by(PlatformConfiguration.version.desc())
    )
    return [PlatformConfigurationRead.model_validate(row) for row in rows.all()]


@router.put("", response_model=PlatformConfigurationRead)
async def update_configuration(
    payload: SaveConfigurationRequest,
    user: Annotated[AuthenticatedUser, config_access],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> PlatformConfigurationRead:
    _check_cross_field_rules(payload)
    async with session.begin():
        previous = await _current(session)
        now = utc_now()
        updated = PlatformConfiguration(
            version=previous.version + 1,
            platform_name=payload.platform_name,
            logo_location=payload.logo_location,
            brand_primary_color=payload.brand_primary_color,
            email_sender_name=payload.email_sender_name,
            email_sender_address=payload.email_sender_address,
            verification_expiration_days=payload.verification_expiration_days,
            supported_countries=payload.supported_countries,
            supported_languages=payload.supported_languages,
            opportunity_categories=payload.opportunity_categories,
            document_types=payload.document_types,
            maximum_file_size_bytes=payload.maximum_file_size_bytes,
            applicant_registration_enabled=payload.applicant_registration_enabled,
            provider_registration_enabled=payload.provider_registration_enabled,
            maintenance_mode=payload.maintenance_mode,
            feature_flags=payload.feature_flags,
            environment=payload.environment,
            security_policy=payload.security_policy,
            recommendation_settings=payload.recommendation_settings,
            fraud_rule_settings=payload.fraud_rule_settings,
            integration_settings=payload.integration_settings,
            notification_settings=payload.notification_settings,
            updated_at=now,
            updated_by=user.uid,
            change_reason=payload.reason,
        )
        session.add(updated)
        await append_audit_record(
            session,
            actor_id=user.uid,
            actor_role=user.role,
            action=AuditAction.administrativeAction,
            entity_type="platform_configuration",
            entity_id="global",
            previous_value=f"version={previous.version}",
            new_value=f"version={updated.version}",
            result=AuditResult.success,
            correlation_id=f"config-{updated.version}",
        )
        await session.flush()
        return PlatformConfigurationRead.model_validate(updated)


@router.post("/rollback", response_model=PlatformConfigurationRead)
async def rollback_configuration(
    payload: RollbackRequest,
    user: Annotated[AuthenticatedUser, config_access],
    session: Annotated[AsyncSession, Depends(get_db)],
) -> PlatformConfigurationRead:
    target = await session.get(PlatformConfiguration, payload.target_version)
    if target is None:
        raise HTTPException(status_code=404, detail="Configuration version was not found.")
    await session.commit()
    restore = SaveConfigurationRequest(
        platform_name=target.platform_name,
        logo_location=target.logo_location,
        brand_primary_color=target.brand_primary_color,
        email_sender_name=target.email_sender_name,
        email_sender_address=target.email_sender_address,
        verification_expiration_days=target.verification_expiration_days,
        supported_countries=target.supported_countries,
        supported_languages=target.supported_languages,
        opportunity_categories=target.opportunity_categories,
        document_types=target.document_types,
        maximum_file_size_bytes=target.maximum_file_size_bytes,
        applicant_registration_enabled=target.applicant_registration_enabled,
        provider_registration_enabled=target.provider_registration_enabled,
        maintenance_mode=target.maintenance_mode,
        feature_flags=target.feature_flags,
        environment=target.environment,
        security_policy=target.security_policy,
        recommendation_settings=target.recommendation_settings,
        fraud_rule_settings=target.fraud_rule_settings,
        integration_settings=target.integration_settings,
        notification_settings=target.notification_settings,
        reason=f"Rollback: {payload.reason}",
    )
    return await update_configuration(restore, user, session)
